File: Algorithm/200518/1449. Form Largest Integer With Digits That Add up to Target.c
#include <stdio.h>

#define N_MAX			9	
#define MAX(x, y)		( (x) > (y) ? (x) : (y) )
int cost[N_MAX];
int N;
int target;

int ans = -1;

void f(int x, int cnt)
{
	if (x == 0) {
		ans = MAX(ans, cnt);
		return;
	}
	for(int i=0; i<N; ++i) {
		if (x < cost[i]) continue;
		f(x-cost[i], cnt+1);
	}
}

int g(int x)
{
	if (x == 0) return 0;
	int ret = -1;
	for(int i=0; i<N; ++i) {
		if (x < cost[i]) continue;
		ret = MAX(ret, g(x-cost[i])+1);
	}
	return ret;


}

int main(int argc, char **argv)
{
	scanf("%d", &N);
	scanf("%d", &target);
	for(int n=0; n<N; ++n) scanf("%d", cost+n);
//	f(target, 0);
//	printf("ans: %d\n", ans);
	printf("%d\n", g(target));
	return 0;
}

File: Algorithm/200518/1449. Form Largest Integer With Digits That Add up to Target.py
'''
	dp[x] = Max(dp[x-cost[i]]*10 + i+1)
'''


class Solution:
    def largestNumber(self, cost: List[int], target: int) -> str:
        dp = [-1]*5001
        dp[0] = 0
        
        for x in range(1, target+1):
            for i,c in enumerate(cost):
                if x < c or dp[x-c] == -1: 
                    continue
                tmp = dp[x-c]*10+i+1
                dp[x] = tmp if dp[x]<tmp else dp[x]
                     
        return "0" if dp[target] == -1 else str(dp[target])
        
		
